File: python/src/c3_chain_of_responsibility.py
"""
The intent is to avoid coupling the sender of a request to its receiver by giving
more than one object a chance to handle the request. Chain the receiving objects
and pass the request along the chain until an object handles it.

Key participants:
- Handler: Defines interface for handling requests and successor link
- ConcreteHandler: Handles requests it's responsible for; accesses successor
- Client: Initiates the request

Key benefits:
- Decouples sender from receiver
- Simplifies object interconnections
- Flexible responsibility assignment

Real-world example: A technical support ticket system with escalation levels.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum
from typing import Optional


# ----- Request -----
class Level(Enum):
    BASIC = "Basic"
    INTERMEDIATE = "Intermediate"
    ADVANCED = "Advanced"
    CRITICAL = "Critical"


@dataclass
class SupportTicket:
    description: str
    level: Level

    def get_level_string(self) -> str:
        return self.level.value


# ----- Handler interface -----
class SupportHandler(ABC):
    def __init__(self) -> None:
        self._next: Optional["SupportHandler"] = None

    def set_next(self, handler: "SupportHandler") -> "SupportHandler":
        self._next = handler
        return handler

    def handle_ticket(self, ticket: SupportTicket) -> None:
        if self._can_handle(ticket):
            self._do_handle(ticket)
        elif self._next:
            print(f"{self.__class__.__name__} escalating...")
            self._next.handle_ticket(ticket)
        else:
            print("No handler available for this ticket.")

    @abstractmethod
    def _can_handle(self, ticket: SupportTicket) -> bool:
        pass

    @abstractmethod
    def _do_handle(self, ticket: SupportTicket) -> None:
        pass


# ----- Concrete Handlers -----
class Level1Support(SupportHandler):
    def _can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.level == Level.BASIC

    def _do_handle(self, ticket: SupportTicket) -> None:
        print(f"[Level 1 Support] Handling: {ticket.description}")
        print("  Issue resolved with basic troubleshooting.")


class Level2Support(SupportHandler):
    def _can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.level == Level.INTERMEDIATE

    def _do_handle(self, ticket: SupportTicket) -> None:
        print(f"[Level 2 Support] Handling: {ticket.description}")
        print("  Issue resolved with technical expertise.")


class Manager(SupportHandler):
    def _can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.level == Level.ADVANCED

    def _do_handle(self, ticket: SupportTicket) -> None:
        print(f"[Manager] Handling: {ticket.description}")
        print("  Issue resolved with management authority.")


class Director(SupportHandler):
    def _can_handle(self, ticket: SupportTicket) -> bool:
        return ticket.level == Level.CRITICAL

    def _do_handle(self, ticket: SupportTicket) -> None:
        print(f"[Director] Handling: {ticket.description}")
        print("  Issue resolved at highest level.")


# ----- Example usage -----
def chain_of_responsibility_example() -> None:
    """
    Demonstrates the Chain of Responsibility pattern with a support ticket system.
    """
    print("\n--- Chain of Responsibility Pattern Example ---")

    # Build the chain
    level1 = Level1Support()
    level2 = Level2Support()
    manager = Manager()
    director = Director()

    level1.set_next(level2).set_next(manager).set_next(director)

    # Create tickets
    tickets = [
        SupportTicket("Can't connect to WiFi", Level.BASIC),
        SupportTicket("Server crash", Level.INTERMEDIATE),
        SupportTicket("Datacenter outage", Level.CRITICAL),
        SupportTicket("Budget approval needed", Level.ADVANCED),
    ]

    for ticket in tickets:
        print(f"\nProcessing: {ticket.description} ({ticket.get_level_string()})")
        level1.handle_ticket(ticket)


# Test helpers
def get_support_chain() -> SupportHandler:
    """Helper for tests: Create a complete support chain."""
    level1 = Level1Support()
    level2 = Level2Support()
    manager = Manager()
    return level1.set_next(level2).set_next(manager)


def get_basic_ticket() -> SupportTicket:
    """Helper for tests: Create a basic level ticket."""
    return SupportTicket("Basic issue", Level.BASIC)


if __name__ == "__main__":
    chain_of_responsibility_example()

File: rust/src/c3_chain_of_responsibility.rs
//! # Chain of Responsibility Pattern
//!
//! The Chain of Responsibility pattern avoids coupling the sender of a request
//! to its receiver by giving more than one object a chance to handle the request.
//!
//! ## Key Participants
//! - **Handler**: Defines interface for handling requests and successor link
//! - **ConcreteHandler**: Handles requests it's responsible for; accesses successor
//! - **Client**: Initiates the request
//!
//! ## Benefits
//! - Decouples sender from receiver
//! - Simplifies object interconnections
//! - Flexible responsibility assignment

use std::cell::RefCell;
use std::rc::Rc;

// Request
#[derive(Debug, Clone, PartialEq)]
pub enum Level {
    Basic,
    Intermediate,
    Advanced,
    Critical,
}

#[derive(Debug, Clone)]
pub struct SupportTicket {
    pub description: String,
    pub level: Level,
}

impl SupportTicket {
    pub fn new(description: &str, level: Level) -> Self {
        Self {
            description: description.to_string(),
            level,
        }
    }

    pub fn get_level_string(&self) -> &str {
        match self.level {
            Level::Basic => "Basic",
            Level::Intermediate => "Intermediate",
            Level::Advanced => "Advanced",
            Level::Critical => "Critical",
        }
    }
}

// Handler trait
pub trait SupportHandler {
    fn set_next(&mut self, handler: Rc<RefCell<dyn SupportHandler>>);
    fn handle_ticket(&self, ticket: &SupportTicket);
    fn can_handle(&self, ticket: &SupportTicket) -> bool;
    fn do_handle(&self, ticket: &SupportTicket);
}

// Concrete Handlers
pub struct Level1Support {
    next: RefCell<Option<Rc<RefCell<dyn SupportHandler>>>>,
}

impl Level1Support {
    pub fn new() -> Self {
        Self {
            next: RefCell::new(None),
        }
    }
}

impl SupportHandler for Level1Support {
    fn set_next(&mut self, handler: Rc<RefCell<dyn SupportHandler>>) {
        *self.next.borrow_mut() = Some(handler);
    }

    fn handle_ticket(&self, ticket: &SupportTicket) {
        if self.can_handle(ticket) {
            self.do_handle(ticket);
        } else if let Some(next) = self.next.borrow().as_ref() {
            println!("Level1Support escalating...");
            next.borrow().handle_ticket(ticket);
        } else {
            println!("No handler available for this ticket.");
        }
    }

    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.level == Level::Basic
    }

    fn do_handle(&self, ticket: &SupportTicket) {
        println!(
            "[Level 1 Support] Handling: {}",
            ticket.description
        );
        println!("  Issue resolved with basic troubleshooting.");
    }
}

pub struct Level2Support {
    next: RefCell<Option<Rc<RefCell<dyn SupportHandler>>>>,
}

impl Level2Support {
    pub fn new() -> Self {
        Self {
            next: RefCell::new(None),
        }
    }
}

impl SupportHandler for Level2Support {
    fn set_next(&mut self, handler: Rc<RefCell<dyn SupportHandler>>) {
        *self.next.borrow_mut() = Some(handler);
    }

    fn handle_ticket(&self, ticket: &SupportTicket) {
        if self.can_handle(ticket) {
            self.do_handle(ticket);
        } else if let Some(next) = self.next.borrow().as_ref() {
            println!("Level2Support escalating...");
            next.borrow().handle_ticket(ticket);
        } else {
            println!("No handler available for this ticket.");
        }
    }

    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.level == Level::Intermediate
    }

    fn do_handle(&self, ticket: &SupportTicket) {
        println!(
            "[Level 2 Support] Handling: {}",
            ticket.description
        );
        println!("  Issue resolved with technical expertise.");
    }
}

pub struct Manager {
    next: RefCell<Option<Rc<RefCell<dyn SupportHandler>>>>,
}

impl Manager {
    pub fn new() -> Self {
        Self {
            next: RefCell::new(None),
        }
    }
}

impl SupportHandler for Manager {
    fn set_next(&mut self, handler: Rc<RefCell<dyn SupportHandler>>) {
        *self.next.borrow_mut() = Some(handler);
    }

    fn handle_ticket(&self, ticket: &SupportTicket) {
        if self.can_handle(ticket) {
            self.do_handle(ticket);
        } else if let Some(next) = self.next.borrow().as_ref() {
            println!("Manager escalating...");
            next.borrow().handle_ticket(ticket);
        } else {
            println!("No handler available for this ticket.");
        }
    }

    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.level == Level::Advanced
    }

    fn do_handle(&self, ticket: &SupportTicket) {
        println!("[Manager] Handling: {}", ticket.description);
        println!("  Issue resolved with management authority.");
    }
}

pub struct Director {
    next: RefCell<Option<Rc<RefCell<dyn SupportHandler>>>>,
}

impl Director {
    pub fn new() -> Self {
        Self {
            next: RefCell::new(None),
        }
    }
}

impl SupportHandler for Director {
    fn set_next(&mut self, handler: Rc<RefCell<dyn SupportHandler>>) {
        *self.next.borrow_mut() = Some(handler);
    }

    fn handle_ticket(&self, ticket: &SupportTicket) {
        if self.can_handle(ticket) {
            self.do_handle(ticket);
        } else if let Some(next) = self.next.borrow().as_ref() {
            println!("Director escalating...");
            next.borrow().handle_ticket(ticket);
        } else {
            println!("No handler available for this ticket.");
        }
    }

    fn can_handle(&self, ticket: &SupportTicket) -> bool {
        ticket.level == Level::Critical
    }

    fn do_handle(&self, ticket: &SupportTicket) {
        println!("[Director] Handling: {}", ticket.description);
        println!("  Issue resolved at highest level.");
    }
}

/// Example demonstrating Chain of Responsibility pattern
pub fn example() {
    println!("\n--- Chain of Responsibility Pattern Example ---\n");

    // Build the chain
    let level1 = Rc::new(RefCell::new(Level1Support::new()));
    let level2 = Rc::new(RefCell::new(Level2Support::new()));
    let manager = Rc::new(RefCell::new(Manager::new()));
    let director = Rc::new(RefCell::new(Director::new()));

    // Set up chain
    level1.borrow_mut().set_next(level2.clone());
    level2.borrow_mut().set_next(manager.clone());
    manager.borrow_mut().set_next(director.clone());

    // Create tickets
    let tickets = vec![
        SupportTicket::new("Can't connect to WiFi", Level::Basic),
        SupportTicket::new("Server crash", Level::Intermediate),
        SupportTicket::new("Datacenter outage", Level::Critical),
        SupportTicket::new("Budget approval needed", Level::Advanced),
    ];

    for ticket in tickets {
        println!(
            "\nProcessing: {} ({})",
            ticket.description,
            ticket.get_level_string()
        );
        level1.borrow().handle_ticket(&ticket);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_ticket_handled_by_level1() {
        let level1 = Rc::new(RefCell::new(Level1Support::new()));
        let level2 = Rc::new(RefCell::new(Level2Support::new()));
        level1.borrow_mut().set_next(level2);

        let ticket = SupportTicket::new("WiFi issue", Level::Basic);
        level1.borrow().handle_ticket(&ticket);
        // Should not panic
    }

    #[test]
    fn test_intermediate_ticket_escalates() {
        let level1 = Rc::new(RefCell::new(Level1Support::new()));
        let level2 = Rc::new(RefCell::new(Level2Support::new()));
        level1.borrow_mut().set_next(level2);

        let ticket = SupportTicket::new("Server crash", Level::Intermediate);
        level1.borrow().handle_ticket(&ticket);
        // Should not panic
    }

    #[test]
    fn test_advanced_ticket_handled_by_manager() {
        let level1 = Rc::new(RefCell::new(Level1Support::new()));
        let manager = Rc::new(RefCell::new(Manager::new()));
        level1.borrow_mut().set_next(manager);

        let ticket = SupportTicket::new("Budget needed", Level::Advanced);
        level1.borrow().handle_ticket(&ticket);
        // Should not panic
    }

    #[test]
    fn test_ticket_get_level_string() {
        let ticket = SupportTicket::new("Test", Level::Basic);
        assert_eq!(ticket.get_level_string(), "Basic");
    }
}
